SYCL/USM/copy.cpp

```diff
@@ -89,39 +118,80 @@ int main() {
   queue q;
   auto dev = q.get_device();
 
-  test_struct test_obj{4, 42, 424, 4242, 4.2f, 4.242f};
+  const bool DoublesSupported = dev.has(sycl::aspect::fp64);
+  const bool HalfsSupported = dev.has(sycl::aspect::fp16);
+
+  test_struct_all test_obj_all{4, 42, 424, 4242, 4.2f, 4.242, 4.24242};
+  test_struct_whalf test_obj_whalf{4, 42, 424, 4242, 4.2f, 4.242};
+  test_struct_wdouble test_obj_wdouble{4, 42, 424, 4242, 4.242, 4.24242};
+  test_struct_minimum test_obj_minimum{4, 42, 424, 4242, 4.242};
 
   if (dev.has(aspect::usm_host_allocations)) {
     runTests<short>(q, 4, alloc::host, alloc::host);
     runTests<int>(q, 42, alloc::host, alloc::host);
     runTests<long>(q, 424, alloc::host, alloc::host);
     runTests<long long>(q, 4242, alloc::host, alloc::host);
-    runTests<half>(q, half(4.2f), alloc::host, alloc::host);
+    if (HalfsSupported)
+      runTests<half>(q, half(4.2f), alloc::host, alloc::host);
     runTests<float>(q, 4.242f, alloc::host, alloc::host);
-    runTests<double>(q, 4.24242, alloc::host, alloc::host);
-    runTests<test_struct>(q, test_obj, alloc::host, alloc::host);
+    if (DoublesSupported)
+      runTests<double>(q, 4.24242, alloc::host, alloc::host);
+    if (HalfsSupported && DoublesSupported)
+      runTests<test_struct_all>(q, test_obj_all, alloc::host, alloc::host);
+    else if (HalfsSupported)
+      runTests<test_struct_whalf>(q, test_obj_whalf, alloc::host, alloc::host);
```

Comment:
Minor comment: I don't see what 'w' stands for in 'test_struct_whalf' and 'test_struct_wdouble', but it seems to be added intentionally, right?

Reply:
I have renamed them to just `test_struct_double` and `test_struct_half`.